[pytorch][triton] Allow warp spec for FlexAttention kernel

mandroid6 opened the pull request:
Summary:
Given inductor support for warp-specialization for `TritonTemplateKernel`, this change adds:
- num_consumer_groups
- num_buffers_warp_spec

to the flexattention template generated by inductor in `torch.compile`.

NOTE: Currently default config doesn't enable warp-spec and needs explicit args for num_consumer_groups, num_buffers_warp_spec in the kernel options to enable.

Test Plan:
### Functional Testing
```Py
import torch
from torch.nn.attention.flex_attention import flex_attention
from triton.testing import do_bench
make_tensor = lambda: torch.rand(8, 16, 8192, 128, device="cuda", dtype=torch.bfloat16)
q, k, v = make_tensor(), make_tensor(), make_tensor()
flex_compiled = torch.compile(flex_attention, fullgraph=True)
print(do_bench(lambda: flex_compiled(q, k, v, kernel_options={"num_warps": 4, "num_consumer_groups": 2,
                "num_buffers_warp_spec": 3,})))
```
- (best config) without WS: 11.06
- with WS: 9.35

Differential Revision: D70501880




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -138,11 +138,7 @@ def maybe_realize(args: list[Optional[IRNode]]):
-        or torch.version.hip
-        or torch.mtia.is_available()
-    ):
+    if torch.get_float32_matmul_precision() == "highest" or torch.version.hip:
```
rebase bug?

mandroid6 replied:
yeah looks like it; updated

drisspg replied:
can you just revert this change entirely, still seems like the MTIA change is gone

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -1124,9 +1120,9 @@ def convert_mask_graph_module(mask_graph):
             "Unsupported for now if query, key, value are the same buffer."
         )
-    if query.get_dtype() not in [torch.float, torch.bfloat16, torch.float16]:
+    if query.get_dtype() not in [torch.float, torch.bfloat16]:
```
same here

drisspg reviewed (approved):
Changes look good, can we add a test w/ a known good config/ size that should work w/ consumer groups

mandroid6 commented:
Added test for num_consumer_groups=2, num_buffers_warp_spec=3

Commit:
[pytorch][triton] Enable warp spec for FlexAttention kernel (#150470)

Summary:

Given inductor support for warp-specialization for `TritonTemplateKernel`, this change adds:
- num_consumer_groups
- num_buffers_warp_spec

to the flexattention template generated by inductor in `torch.compile`.

NOTE: Currently default config doesn't enable warp-spec and needs explicit args for num_consumer_groups, num_buffers_warp_spec in the kernel options to enable.

Test Plan:
### Functional Testing
```
import torch
from torch.nn.attention.flex_attention import flex_attention
from triton.testing import do_bench
make_tensor = lambda: torch.rand(8, 16, 8192, 128, device="cuda", dtype=torch.bfloat16)
q, k, v = make_tensor(), make_tensor(), make_tensor()
flex_compiled = torch.compile(flex_attention, fullgraph=True)
print(do_bench(lambda: flex_compiled(q, k, v, kernel_options={"num_warps": 4, "num_consumer_groups": 2,
                "num_buffers_warp_spec": 3,})))
```
- (best config) without WS: 11.06
- with WS: 9.35

Reviewed By: htyu

Differential Revision: D70501880

Camyll commented:
@pytorchbot merge -i
